### PR TITLE
Highlight diff

### PR DESCRIPTION
**Pull request by andreynering**
cc @bkcsoft

Following up https://github.com/gogits/gogs/pull/2296

Not all lines are being highlighted, but not even GitHub or GitLab do it.

### CODE REVIEW

**Comment by bkcsoft** [comment]
I'll have a look later on but I like the idea :)

**Comment by Unknwon** [comment]
Thanks! Will review after #2296

**Comment by bkcsoft** [comment]
@andreynering ypu'll have to rebase on #2296 again :)

**Comment by andreynering** [comment]
@Unknwon  @bkcsoft 

I rebased with `develop`, improved my code and just updated the pull request.

Here's a picture for who's curious:

**Comment by Unknwon** [comment]
Awesome, I'll checkout locally and test.

File: models/git_diff.go [hunk 1]
```diff
@@ -23,32 +25,37 @@ import (
 	"github.com/gogits/gogs/modules/base"
 	"github.com/gogits/gogs/modules/log"
 	"github.com/gogits/gogs/modules/process"
+	"github.com/sergi/go-diff/diffmatchpatch"
```
Comment (Unknwon) [inline models/git_diff.go]: Please move this line after line 20.

File: models/git_diff.go [hunk 2]
```diff
@@ -23,32 +25,37 @@ import (
...
+
 const (
-	DIFF_LINE_PLAIN = iota + 1
-	DIFF_LINE_ADD
-	DIFF_LINE_DEL
-	DIFF_LINE_SECTION
+	DIFF_LINE_PLAIN   DiffLineType = iota + 1
+	DIFF_LINE_ADD     DiffLineType = iota + 1
+	DIFF_LINE_DEL     DiffLineType = iota + 1
+	DIFF_LINE_SECTION DiffLineType = iota + 1
```
Comment (Unknwon) [inline models/git_diff.go]: `DiffLineType = iota + 1` is redundant since line 35.

File: models/git_diff.go [hunk 3]
```diff
@@ -23,32 +25,37 @@ import (
...
+
 const (
-	DIFF_FILE_ADD = iota + 1
-	DIFF_FILE_CHANGE
-	DIFF_FILE_DEL
-	DIFF_FILE_RENAME
+	DIFF_FILE_ADD    DiffFileType = iota + 1
+	DIFF_FILE_CHANGE DiffFileType = iota + 1
+	DIFF_FILE_DEL    DiffFileType = iota + 1
+	DIFF_FILE_RENAME DiffFileType = iota + 1
```
Comment (Unknwon) [inline models/git_diff.go]: `DiffFileType = iota + 1` is redundant since line 44.

**Comment by andreynering** [comment]
@Unknwon I did the fixes.

File: models/git_diff.go [hunk 4]
```diff
@@ -56,12 +63,93 @@ type DiffSection struct {
...
+		} else if s.Type == diffmatchpatch.DiffDelete && lineType == DIFF_LINE_DEL {
+			result = result + "<span class=\"removed-code\">"+html.EscapeString(s.Text)+"</span>"
+		} else if s.Type == diffmatchpatch.DiffEqual {
+			result = result + html.EscapeString(s.Text)
+		}
+	}
+	return template.HTML(result)
+}
+
+func (diffSection *DiffSection) GetLeftLine(idx int, sliceIdx int) *DiffLine {
```
Comment (Unknwon) [inline models/git_diff.go]: Can you explain what this does?

Comment (andreynering) [inline models/git_diff.go]: Sure.

To [compute the diff](https://github.com/andreynering/gogs/blob/bf11ad19eab0de5aef297a21ed58c6cd811a137c/models/git_diff.go#L140) I must have two line strings, the deleted and the added one. `GetLeftLine` try to get the deleted one when the actual is an add. `GetRightLine` is the oposite. [See](https://github.com/andreynering/gogs/blob/bf11ad19eab0de5aef297a21ed58c6cd811a137c/models/git_diff.go#L122-L137).

Unfortunaly, this isn't working when the left index is different of the right index. [Example](https://github.com/andreynering/gogs/commit/81ed5c4bee112beacd8e0f34b336b4f3b6c6a6c1#diff-478b0554d47378204e26103fd20ec06bL33): left is 33 and right 34 for the same line.

I tried to fix that, but it's tricky. Any help would be appreciated.

**Comment by andreynering** [comment]
I added another commit to the PR. I think I finally got it working now. :grinning:

**Comment by Unknwon** [comment]
LGTM, I'll merge now but will have comment for questions.

**Pull request merged**